File: windows/particles/OpenBurnBar.Particles/Substrates/Catalog/SubstrateCatalog.cs
using System.Collections.Generic;
using System.Linq;
using OpenBurnBar.Particles.Model;

namespace OpenBurnBar.Particles.Substrates.Catalog;

/// <summary>
/// The single substrate registry — C# port of Swift
/// <c>Views/Substrate/SubstrateCatalog.swift</c>. Aggregates one shared "Plain · DOTS"
/// descriptor per family plus each family's bespoke painters. Family registries are the
/// only per-family edit points (matching the Swift design), so each ported family
/// (Constellation, Volumetric here; Flow / Aurora / Mesh / Moire as those lanes land)
/// appends its own <c>&lt;Family&gt;Family.Styles</c> array below — this file is the one
/// documented merge point.
/// </summary>
public static class SubstrateCatalog
{
    public const string PlainId = "plain";

    /// <summary>The families whose bespoke painters are registered in the C# port so far.</summary>
    public static readonly IReadOnlyList<SubstrateFamily> RegisteredFamilies = new[]
    {
        SubstrateFamily.Constellation,
        SubstrateFamily.Volumetric,
    };

    /// <summary>Plain descriptors (one per registered family; all reuse <see cref="PlainDotsSubstrate"/>).</summary>
    private static readonly IReadOnlyList<SubstrateDescriptor> PlainDescriptors =
        RegisteredFamilies.Select(fam => new SubstrateDescriptor(
            PlainId, fam, "Plain", "dots",
            FamilyAccent.A(fam), FamilyAccent.A2(fam),
            static () => new PlainDotsSubstrate())).ToArray();

    /// <summary>Plain + every bespoke descriptor across the registered families.</summary>
    public static readonly IReadOnlyList<SubstrateDescriptor> SubstrateList =
        PlainDescriptors
            .Concat(ConstellationFamily.Styles)
            .Concat(VolumetricFamily.Styles)
            .ToArray();

    /// <summary>Lookup by id. The plain descriptors share id <c>"plain"</c>; the first wins.</summary>
    public static readonly IReadOnlyDictionary<string, SubstrateDescriptor> ById =
        SubstrateList
            .GroupBy(d => d.Id)
            .ToDictionary(g => g.Key, g => g.First());

    /// <summary>The bespoke descriptors for one family (no plain).</summary>
    public static IEnumerable<SubstrateDescriptor> Bespoke(SubstrateFamily fam) =>
        SubstrateList.Where(d => d.Family == fam && d.Id != PlainId);

    /// <summary>The shared plain descriptor for one family.</summary>
    public static SubstrateDescriptor Plain(SubstrateFamily fam) =>
        PlainDescriptors.First(d => d.Family == fam);

    /// <summary>A family's full offering — its plain first, then its bespoke painters.</summary>
    public static IEnumerable<SubstrateDescriptor> Entries(SubstrateFamily fam) =>
        new[] { Plain(fam) }.Concat(Bespoke(fam));
}

File: OpenBurnBarCore/Sources/OpenBurnBarUI/Views/Substrate/SubstrateCatalog.swift
import SwiftUI
import OpenBurnBarKernel

// MARK: - Substrate descriptor + catalog

/// Static metadata for one substrate style + a factory to build its renderer.
public struct SubstrateDescriptor: Identifiable, Sendable {
    /// Globally-unique id: `"plain"` or `"<family>.<style>"` (persisted).
    public let id: String
    public let family: SubstrateFamily
    public let label: String          // "Stellar Plasma"
    public let hint: String           // ONE texture word: "twinkle"
    public let accent: RGBA           // picker tile accent a
    public let accent2: RGBA          // picker tile accent a2
    /// Builds a fresh renderer instance (cached & reused by the simulation).
    /// `@Sendable` (the factories capture nothing) so descriptors are `Sendable`
    /// and the static catalog lists are concurrency-safe.
    public let make: @MainActor @Sendable () -> SwarmSubstrate

    public init(id: String, family: SubstrateFamily, label: String, hint: String,
                accent: RGBA, accent2: RGBA, make: @escaping @MainActor @Sendable () -> SwarmSubstrate) {
        self.id = id; self.family = family; self.label = label; self.hint = hint
        self.accent = accent; self.accent2 = accent2; self.make = make
    }
}

/// The single substrate registry. The bespoke authors never touch this file —
/// each only edits their own family registry (`Families/<Family>Family.swift`).
/// This aggregates the six family arrays + the one shared plain, giving the
/// 30-entry catalog exactly seven edit points.
public enum SubstrateCatalog {
    public static let plainID = "plain"

    /// All 30 = 6 plain descriptors (one shared renderer) + 24 bespoke.
    public static let substrateList: [SubstrateDescriptor] =
        plainDescriptors
        + ConstellationFamily.styles
        + FlowFamily.styles
        + AuroraFamily.styles
        + MeshFamily.styles
        + MoireFamily.styles
        + VolumetricFamily.styles

    /// Lookup by id. All six plain descriptors share id `"plain"`; the first wins
    /// (they're the same family-agnostic renderer, so it doesn't matter).
    public static let byID: [String: SubstrateDescriptor] =
        Dictionary(substrateList.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })

    /// The styles offered for the active kernel — its family's five, plain first.
    public static func styles(forKernel kernelID: String) -> [SubstrateDescriptor] {
        let fam = SubstrateFamily.forKernel(kernelID)
        return [plain(for: fam)] + bespoke(for: fam)
    }

    public static func entries(in family: SubstrateFamily) -> [SubstrateDescriptor] {
        [plain(for: family)] + bespoke(for: family)
    }

    /// Resolve the descriptor to actually render. The selected id only "lights up"
    /// on the kernel theme whose family owns it; otherwise the active family's
    /// plain is used — exactly the web behavior where switching worlds shows that
    /// world's suite.
    public static func resolved(forKernel kernelID: String, selectedID: String) -> SubstrateDescriptor {
        let fam = SubstrateFamily.forKernel(kernelID)
        if let d = byID[selectedID], d.family == fam { return d }
        if selectedID == plainID { return plain(for: fam) }
        return plain(for: fam)
    }

    /// Convenience: the resolved descriptor for the live `UserDefaults` selection.
    public static func resolvedCurrent(forKernel kernelID: String) -> SubstrateDescriptor {
        guard SwarmSubstratePreferences.isEnabled else { return plain(for: .forKernel(kernelID)) }
        return resolved(forKernel: kernelID, selectedID: SwarmSubstratePreferences.currentID)
    }

    // MARK: Internals

    /// One "Plain · DOTS" descriptor per family (all reuse `PlainDotsSubstrate`).
    private static let plainDescriptors: [SubstrateDescriptor] =
        SubstrateFamily.allCases.map { fam in
            SubstrateDescriptor(
                id: plainID, family: fam, label: "Plain", hint: "dots",
                accent: FamilyAccent.a(fam), accent2: FamilyAccent.a2(fam),
                make: { PlainDotsSubstrate() })
        }

    public static func plain(for fam: SubstrateFamily) -> SubstrateDescriptor {
        plainDescriptors.first { $0.family == fam }!
    }

    public static func bespoke(for fam: SubstrateFamily) -> [SubstrateDescriptor] {
        substrateList.filter { $0.family == fam && $0.id != plainID }
    }
}
